wallet/src/state.rs

```diff
@@ -188,12 +227,168 @@ impl ClientState {
         }
     }
 
-    /// Generate a new transaction.
-    ///
-    /// TODO: this function is too complicated, merge with
-    /// builder API ?
+    /// Generate a new transaction delegating stake
+    #[instrument(skip(self, rng))]
+    pub fn build_delegate<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        rate_data: RateData,
+        unbonded_amount: u64,
+        fee: u64,
+        source_address: Option<u64>,
+    ) -> Result<Transaction, anyhow::Error> {
+        // If the source address is set, send the delegation tokens to the same
+        // address; otherwise, send them to the default address.
+        let (_label, self_address) = self
+            .wallet()
+            .address_by_index(source_address.unwrap_or(0) as usize)?;
+
+        let mut tx_builder = Transaction::build_with_root(self.note_commitment_tree.root2());
+
+        tx_builder
+            .set_fee(fee)
+            // XXX populate chain id from params?
+            .set_chain_id(CURRENT_CHAIN_ID.to_string())
```

Comment:
Now that we have `ChainParams`, we should look through all calls to `.set_chain_id` and have them pass a string from the chain params, rather than a hardcoded constant.